### PR TITLE
Extend coding guidelines

### PR DESCRIPTION
**Pull request by TimSatke**

### CODE REVIEW

File: CODING_GUIDELINES.md [hunk 1]
```diff
@@ -1,13 +1,140 @@
...
+if err == somepkg.ErrTimeout {
+    // timeout, maybe try again?
+} else if err != nil {
+    // some other error, handle accordingly
+}
+```
+If you make use of sentinel errors in your package, **sentinel errors must be constant**.
+Go itself doesn't follow this rule.
+As an example, `io.EOF` is not a constant error.
+Run a larger application after calling `io.EOF = nil` and see what happens.
```
Comment (SUMUKHA-PK) [inline CODING_GUIDELINES.md:92]: Was this supposed to be `io.EOF == nil` ?

File: CODING_GUIDELINES.md [hunk 2]
```diff
@@ -1,13 +1,140 @@
...
+6. struct declarations (followed by constructor functions)
+7. exported methods
+8. unexported methods
+9. exported functions
+10. unexported functions
+
+If there are multiple independent structs in one file, you should probably split the file into two or more.
+
+### Constructor functions
+Use the `&struct` syntax for initialization whenever possible.
```
Comment (Abby3017) [inline CODING_GUIDELINES.md:46]: why pointer initialisation made mandatory ??

File: CODING_GUIDELINES.md [hunk 3]
```diff
@@ -16,11 +143,30 @@ After finishing your work, create a PR.
 If you want early reviews and feedback, create a draft PR.
 See the codeowners file, to see what reviewers make sense.
 
+Before committing, run `make lint test` on your system.
+If that fails, the CI build will probably fail as well.
+If `make lint test` doesn't fail on your machine but in CI, that's even worse.
```
Comment (Abby3017) [inline CODING_GUIDELINES.md:148]: good one..  👍

File: CODING_GUIDELINES.md (hunk 2 above)
Comment (SUMUKHA-PK) [reply CODING_GUIDELINES.md:46]: Constructors always need to return the pointer instead of the variable.

Comment (Abby3017) [reply CODING_GUIDELINES.md:46]: If we are not using constructor, like using variable in a fuction there we can use as variable ?
Why constructor need to always return the pointer reasons for that ?

File: CODING_GUIDELINES.md (hunk 1 above)
Comment (TimSatke) [reply CODING_GUIDELINES.md:92]: No. `io.EOF` is not constant, and you can assign `nil` to it.

File: CODING_GUIDELINES.md (hunk 2 above)
Comment (TimSatke) [reply CODING_GUIDELINES.md:46]: Constructors are usually named `NewXXX`, and should return a pointer, just as `new(T)` returns a pointer. `make(T)` does not.

Comment (TimSatke) [reply CODING_GUIDELINES.md:46]: When you create a variable somewhere else, you can use a different method, if it makes sense. This paragraph only regards constructor functions.

**Pull request merged**